pallets/dca/src/lib.rs

```diff
@@ -0,0 +1,995 @@
+		From<<T as pallet::Config>::Asset>,
+
+	<<T as pallet::Config>::Currency as orml_traits::MultiCurrency<<T as frame_system::Config>::AccountId>>::Balance:
+		From<u128>,
+{
+	fn ensure_next_blocknumber_is_bigger_than_current_block(
+		next_execution_block: Option<T::BlockNumber>,
+	) -> DispatchResult {
+		if let Some(next_exection_block) = next_execution_block {
+			let current_block_number = frame_system::Pallet::<T>::current_block_number();
+			ensure!(
+				next_exection_block > current_block_number,
+				Error::<T>::BlockNumberIsNotInFuture
+			);
+		};
+
+		Ok(())
+	}
+
+	fn ensure_total_amount_is_bigger_than_storage_bond(
+		schedule: &Schedule<T::AccountId, T::Asset, T::BlockNumber>,
+	) -> DispatchResult {
+		let min_total_amount = if schedule.order.get_asset_in() == T::NativeAssetId::get() {
+			T::StorageBondInNativeCurrency::get()
+		} else {
+			Self::get_storage_bond_in_sold_currency(&schedule.order)?
+		};
+
+		ensure!(
+			schedule.total_amount > min_total_amount,
+			Error::<T>::TotalAmountShouldBeLargerThanStorageBond
+		);
+
+		Ok(())
+	}
+
+	fn ensure_schedule_is_suspended(schedule_id: ScheduleId) -> DispatchResult {
+		ensure!(
+			Suspended::<T>::contains_key(schedule_id),
+			Error::<T>::ScheduleMustBeSuspended
+		);
+
+		Ok(())
+	}
+
+	fn ensure_schedule_exists(schedule_id: &ScheduleId) -> DispatchResult {
+		ensure!(Schedules::<T>::contains_key(schedule_id), Error::<T>::ScheduleNotExist);
+
+		Ok(())
+	}
+
+	fn ensure_origin_is_schedule_owner(schedule_id: ScheduleId, who: &T::AccountId) -> DispatchResult {
+		let schedule = Schedules::<T>::get(schedule_id).ok_or(Error::<T>::ScheduleNotExist)?;
+		ensure!(*who == schedule.owner, Error::<T>::Forbidden);
+
+		Ok(())
+	}
+
+	fn ensure_sell_amount_is_bigger_than_transaction_fee(
+		schedule: &Schedule<T::AccountId, T::Asset, T::BlockNumber>,
+	) -> DispatchResult {
+		match schedule.order {
+			Order::Sell {
+				asset_in, amount_in, ..
+			} => {
+				let transaction_fee = Self::get_transaction_fee(asset_in)?;
+				ensure!(amount_in > transaction_fee, Error::<T>::TradeAmountIsLessThanFee);
+			}
+			Order::Buy { .. } => {
+				//For buy we don't check as the calculated amount in will always include the fee
+			}
+		}
+
+		Ok(())
+	}
+
+	pub fn execute_schedule(current_blocknumber: T::BlockNumber, weight: &mut u64, schedule_id: ScheduleId) {
+		*weight += Self::get_execute_schedule_weight();
+
+		let schedule = exec_or_return_if_none!(Schedules::<T>::get(schedule_id));
+		let origin: OriginFor<T> = Origin::<T>::Signed(schedule.owner.clone()).into();
+		let blocknumber_for_schedule = exec_or_return_if_none!(current_blocknumber.checked_add(&schedule.period));
+
+		let sold_currency = schedule.order.get_asset_in();
+		if exec_or_return_if_err!(Self::price_change_is_bigger_than_max_allowed(
+			sold_currency,
+			schedule.order.get_asset_out()
+		)) {
+			exec_or_return_if_err!(Self::plan_schedule_for_block(blocknumber_for_schedule, schedule_id));
+			return;
+		}
+
```

Comment:
Using those macros, imo it is really hard to understand what is going on, at least for me. 

in this statement, you have exec or return twice, and then you still return anyway. 

i am not really fan of those macros, becuase even if the naming is more or less clear, it hides the fact that the execution can stop there.  and then when used in IF statement, it is bit of a mind exercise.

Reply:
good point, removed the macro